In [186]:
# Import libraries

import pandas as pd
import numpy as np
import os

## Step 1. Create the “price_label” and “busiest_day” columns.

In [189]:
# Create path variable for re-use when importing data
path = r'C:\Users\sorca\Documents\Achievement_4\4_7\Task 4_7'

In [191]:
# Load 'orders_products_prior.csv' file
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [192]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [197]:
# Verify if the number of rows/columns loaded in the dataframe is the same as the expected numbers.
df_ords_prods_merge.shape

(32399732, 14)

In [199]:
# The dataframe was loaded with the correct data

In [201]:
# Create a dataframe which has the first 1 million rows from 'df_ords_prods_merge'.
df_ords_prods_merge_1mil = df_ords_prods_merge[:1000000]

In [203]:
# Check the shape of the new dataframe
df_ords_prods_merge_1mil.shape

(1000000, 14)

### Step 1.a. Create 'price_label' column using a user defined function

In [206]:
# Create 'price_label' function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [208]:
# Apply the function 'price_label' on every row in the 'df_ords_prods_merge_1mil' dataframe.
df_ords_prods_merge_1mil['price_range'] = df_ords_prods_merge_1mil.apply(price_label, axis=1)

C:\Users\sorca\AppData\Local\Temp\ipykernel_28372\569618643.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ords_prods_merge_1mil['price_range'] = df_ords_prods_merge_1mil.apply(price_label, axis=1)


In [25]:
df_ords_prods_merge_1mil['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673293
Low-range product    314432
High range            12275
Name: count, dtype: int64

In [27]:
# The results above are different than the ones showing in the Exercise - if the 1 million rows are selected randomly,
# then my subset is different from the one in the Exercise.
# Wanted to see if there are products in my subset priced above $15
df_high_priced = df_ords_prods_merge_1mil[df_ords_prods_merge_1mil['prices'] > 15]
df_high_priced

order_id  user_id  order_number  orders_day_of_week  \
398      2086598        6             1                   5   
401       298250        6             2                   4   
436      2402008        7             2                   1   
465       121053        7             3                   0   
480      1695742        7             4                   2   
...          ...      ...           ...                 ...   
999949   2177312     6475             9                   3   
999951   2942296     6475            10                   4   
999953   2942296     6475            10                   4   
999954   2942296     6475            10                   4   
999965    522662     6475            12                   5   

        order_hour_of_day  days_since_prior_order  product_id  \
398                    18                     NaN       38293   
401                    16                     6.0       38293   
436                    19                    30.0       13198   
465                    18                    30.0       13198   
480                    10                     9.0       13198   
...                   ...                     ...         ...   
999949                  9                    13.0       15599   
999951                  1                    15.0       15599   
999953                  1                    15.0       35994   
999954                  1                    15.0       25420   
999965                 22                    13.0       15599   

        add_to_cart_order  reordered  \
398                     2          0   
401                     1          1   
436                    14          0   
465                    22          1   
480                    13          1   
...                   ...        ...   
999949                  6          1   
999951                  2          1   
999953                  4          0   
999954                  5          0   
999965                  1          1   

                                    product_name  aisle_id  department_id  \
398                         Ground Turkey Breast        35             12   
401                         Ground Turkey Breast        35             12   
436                         85% Lean Ground Beef       122             12   
465                         85% Lean Ground Beef       122             12   
480                         85% Lean Ground Beef       122             12   
...                                          ...       ...            ...   
999949  Naturally Hickory Smoked Thick Cut Bacon       106             12   
999951  Naturally Hickory Smoked Thick Cut Bacon       106             12   
999953                              Turkey Wings        35             12   
999954                       Smoked Turkey Necks        35             12   
999965  Naturally Hickory Smoked Thick Cut Bacon       106             12   

        prices _merge price_range  
398       19.6   both  High range  
401       19.6   both  High range  
436       23.2   both  High range  
465       23.2   both  High range  
480       23.2   both  High range  
...        ...    ...         ...  
999949    16.7   both  High range  
999951    16.7   both  High range  
999953    24.0   both  High range  
999954    17.0   both  High range  
999965    16.7   both  High range  

[12275 rows x 15 columns]

In [33]:
# Also wanted to see which is the highest price in my subset
df_ords_prods_merge_1mil['prices'].max()

25.0

In [ ]:
# Although different from the one in the Exercise, it seems to me that the data in my subset is correct.

### Step 1.b. Create 'price_label' column using a the loc() function

In [40]:
# Create conditions using the loc() function
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [42]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [44]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [46]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

In [ ]:
# The number of the items labeled as 'High-range product' is different than the one displayed in the Exercise (412551 vs 417678).
# The sum of the numbers above is 32,399,732 which matches the number of rows of the 'df_ords_prods_merge' immediately after importing,
# which demonstrates that the dataframe 'df_ords_prods_merge' has data which is correct.

### Step 1.c. Create and populate the 'busiest day' column 

In [59]:
# Calculate the frequency for the 'orders_day_of_week' ; will consider 0 as Monday, 6 as Sunday, and the others in between 0 and 6.
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6203329
1    5659298
6    4495887
2    4213105
5    4205076
3    3839865
4    3783172
Name: count, dtype: int64

In [65]:
# Create a list with the 3 values below, based on the day of the week
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [71]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [74]:
# Assign the values in the result list to a new column, named 'busiest day', in the 'df_ords_prods_merge' dataframe.
df_ords_prods_merge['busiest day']= result

In [76]:
# Check the frequency of the values for the 'busiest day' column.
df_ords_prods_merge['busiest day'].value_counts(dropna = False)

busiest day
Regularly busy    22413231
Busiest day        6203329
Least busy         3783172
Name: count, dtype: int64

In [ ]:
# Verificatin: the numbers above add up to the original number of rows when the 'df_ords_prods_merge' dataframe was imported.

## Step 2. Create a new column to house the values 'Busiest Days' and 'Least busy', which will have a different meaning.

In [136]:
# Create a new list and populate it according to the new requirements
result_1 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_1.append("Busiest days")
  elif value == 4 or value == 3:
    result_1.append("Slowest days")
  else:
    result_1.append("Regularly busy")

In [138]:
# Verify the contents of the new list
result_1

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest day

In [140]:
# Assign the values in the result list to a new column, named 'busiest days', in the 'df_ords_prods_merge' dataframe.
df_ords_prods_merge['busiest days']= result_1

In [144]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge    price_range_loc     busiest day    busiest days  
0   both  Mid-range product  Regularly busy  Regularly busy  
1   both  Mid-range product  Regularly busy  Regularly busy  
2   both  Low-range product  Regularly busy  Regularly busy  
3   both  Low-range product  Regularly busy  Regularly busy  
4   both  Low-range product  Regularly busy  Regularly busy

In [146]:
df_ords_prods_merge.shape

(32399732, 17)

## Step 3. Check the values of this new column for accuracy. 

In [149]:
# Check the frequency of the values for the 'busiest day' column.
df_ords_prods_merge['busiest days'].value_counts(dropna = False)

busiest days
Regularly busy    12914068
Busiest days      11862627
Slowest days       7623037
Name: count, dtype: int64

#### Observations:

##### The sum of the count numbers above (for Regularly busy, Busiest days, Slowest days) is 32,399,732, which matchesthe number of rows of the initial number of rows of the 'df_ords_prods_merge' dataframe.

##### The sum of the orders for Monday (0) and Tuesday (1) matches the 'Busiest days' count: 6203329 + 5659298 = 11,862,627.

##### Also, the sum of the orders for Thursday (3) and Friday (4) match the 'Slowest days' count: 3839865 + 3783172 = 7,623,037.

##### Finally, the sum of the orders of the other days of the week match 'Regularly busy' count = 12,914,068.

## Step 4. Create a new column called “busiest_period_of_day.”

In [156]:
# Check the frequency of the values for the 'order_hour_of_day' column.
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761333
11    2735694
14    2688728
15    2661718
13    2660570
12    2618104
16    2534744
9     2453842
17    2087273
8     1717863
18    1636226
19    1258076
20     976000
7      890923
21     795528
22     634159
23     402272
6      290450
0      218742
1      115683
5       87944
2       69360
4       53232
3       51268
Name: count, dtype: int64

In [160]:
# Create a new list and populate it according to the requirements:
result_2 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in (10, 11, 14, 15, 13, 12, 16, 9):
    result_2.append("Most orders")
  elif value in (17, 8, 18, 19, 20, 7, 21, 22):
    result_2.append("Average orders")
  else:
    result_2.append("Fewest orders")

In [162]:
result_2

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [164]:
# Assign the values in the result_1 list to a new column, named 'busiest_period_of_day', in the 'df_ords_prods_merge' dataframe.
df_ords_prods_merge['busiest_period_of_day']= result_2

In [166]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge    price_range_loc     busiest day    busiest days  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy  Regularly busy   
2   both  Low-range product  Regularly busy  Regularly busy   
3   both  Low-range product  Regularly busy  Regularly busy   
4   both  Low-range product  Regularly busy  Regularly busy   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders

## Step 5. Print the frequency for the 'busiest_period_of_day' column.

In [169]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21114733
Average orders     9996048
Fewest orders      1288951
Name: count, dtype: int64

## Step 7. Export 'df_ords_prods_merge' dataframe as a pickle file.

In [172]:
# Verify the number of rows in 'df_ords_prods_merge'
df_ords_prods_merge.shape

(32399732, 18)

In [174]:
# The number of rows matches the initial number of rows, when 'df_ords_prods_merge' was imported.

In [176]:
# Export dataframe to pickle file
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_4_7.pkl'))

In [180]:
# Verify if the dataframe was exported correctly
df_ords_prods_merge_import = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4_7.pkl'))

In [182]:
df_ords_prods_merge_import

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32399727   2977660   206209            13                   1   
32399728   2977660   206209            13                   1   
32399729   2977660   206209            13                   1   
32399730   2977660   206209            13                   1   
32399731   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32399727                 12                     7.0       14197   
32399728                 12                     7.0       38730   
32399729                 12                     7.0       31477   
32399730                 12                     7.0        6567   
32399731                 12                     7.0       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32399727                  5          1   
32399728                  6          0   
32399729                  7          0   
32399730                  8          0   
32399731                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32399727                             Tomato Paste         9              9   
32399728          Brownie Crunch High Protein Bar         3             19   
32399729    High Protein Bar Chunky Peanut Butter         3             19   
32399730      Chocolate Peanut Butter Protein Bar         3             19   
32399731      Roasted & Salted Shelled Pistachios       117             19   

          prices _merge    price_range_loc     busiest day    busiest days  \
0            9.0   both  Mid-range product  Regularly busy  Regularly busy   
1           12.5   both  Mid-range product  Regularly busy  Regularly busy   
2            4.4   both  Low-range product  Regularly busy  Regularly busy   
3            4.7   both  Low-range product  Regularly busy  Regularly busy   
4            1.0   both  Low-range product  Regularly busy  Regularly busy   
...          ...    ...                ...             ...             ...   
32399727     5.6   both  Mid-range product  Regularly busy    Busiest days   
32399728     5.9   both  Mid-range product  Regularly busy    Busiest days   
32399729     4.2   both  Low-range product  Regularly busy    Busiest days   
32399730     4.9   both  Low-range product  Regularly busy    Busiest days   
32399731     1.7   both  Low-range product  Regularly busy    Busiest days   

         busiest_period_of_day  
0           

In [184]:
# It is confirmed that the pickle file has the correct data.